### ISSUE TITLE
Feature: Dark Mode GUI prototype

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#921]
Starting a thread for ongoing work on Dark Mode GUI (see #805 for the initial discussion and steps around this).

**Comment by mpogue2** [comment]
Commit 2dde5267994d944be2c563bab0319e751b4cd8e0 adds a first cut at a waveformSlider. Here's what it looks like right now:

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/3a440e95-8b03-416c-8653-5b7fd3304a24)

NOTE: The grey waveform is faked right now (just random numbers), but the green loop markers and blue current position marker are functional.  Right now it's just one way:  mouse used in seekBar can affect the darkSeekBar, but not vise-versa.  More coming.

See also feature #79 that's been open for quite a while.  I think that moving to QGraphicsItem's allows us to do some very nice stuff (for example, I've darkened the part of waveform that is to the left of the left loop marker, and that which is to the right of the right loop marker.  This is done by overlaying a rectangle with transparency, and letting the QGraphicsScene do all the work (it's pretty efficient, and fairly easy to write code for).

**Comment by mpogue2** [comment]
TODO:
- singing call coloring
- gang the darkSeekBar back to the seekBar
- allow width to change as window size is changed
- figure out how to switch back/forth between Light and Dark modes
- implement darkSongTable (it's a placeholder right now)
- implement new search box (similar to Mixxx)
- implement treeView control (it's a placeholder right now), leveraging search box
- mechanism for customizing colors (external yaml file?  CSS file?  CSV file?) -- already have a way for expert users to customize SVG elements
- clock coloring
- warningLabel -- where?
- allow window resizing

**Comment by mpogue2** [comment]
Sample of waveform coloring for singing calls, as of commit: 5cabb091b7571acb9f14f6683e6d014f99998c38

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/8888a38f-d9a2-4e2b-ad60-0cd64316b087)

The waveform data is still fake, so this is just a prototype of being able to color the waveform for singing calls.  I also tweaked the position indicator to add little triangles at top and bottom, like the Mixxx style.

And, there's some clock coloring now.  The Clear Clock Coloring context menu works too.

**Comment by mpogue2** [comment]
Sample of waveform coloring for patter (with loop points):

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/44543259-b2f8-4411-9eaf-94ee81c30046)

**Comment by mpogue2** [comment]
TODO:
- Bug: current position indicator slides just under the right edge of the waveform slider frame.

**Comment by mpogue2** [comment]
Commit: 

Changing the Music tab's seekBar or the DarkMode tab's darkSeekBar will change the other accordingly.  Note that the units of the seekBar are 0 to songLength-1, while the units of the darkSeekBar are always 0 to 491.  

NOTE: to interrupt loops of slider changes, the other slider is only updated if there is a 1-unit difference from where it is now from where it should be.  This should work regardless of the maximum() of each slider (which determines the number of units in that slider).

**Comment by mpogue2** [comment]
Commit: 128407ecadafd356bae62d7e251450d9044ea6c8
Clock coloring layering is correct now.  Clear Clock Coloring happens faster now.  Tweaked hand lengths and clock coloring arc size.

**Comment by mpogue2** [comment]
**TODO:**
- make warningLabelDark clickable to reset (like warningLabel, descend from clickableLabel)
- elision for the songTitle field?
- Bug?: LONG TIP mode doesn't turn off on STOP

**Comment by mpogue2** [comment]
Commit: 5bb167d9f42cfb91e0ba8c46ecb4ef1259fe3d46

Moved around the time-in-song a bit.  Modified stylesheets so that flash calls would not change size of the song title.  Added tooltips to the time labels.  Adjusted colors of the warningLabel for Dark Mode.

NOTE: the DARKMODE #define is in two places right now....

**Comment by mpogue2** [comment]
Current look (with the darkWarningLabel stuck into a likely location):

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/02f9019d-3927-4821-8836-c85971fb342a)

The warningLabel is used to show time-in-tip and singing-call-section and warning messages like "LONG TIP".  I think this location will do for now.

**Comment by mpogue2** [comment]
STILL TODO:
- a real waveform, that is swapped out/in when a new song is loaded, and colored by patter vs singing call (can we use the fact that the store of the bg pointer is atomic? double-check this...)
- darkSearch field (have it set the search fields on the Music tab for now)
- TreeView (have it set the darkSearch field for now).  Can we use songTable customized, or does it have to be a new thing?  Does it have to swap out/in to support the iTunes playlists and History (or just delete these for now, and make them future features...cool, but not critical).
- darkSongTable (can it just be a regular songTable but styled?  Or to support column selection dropdown does it need to be a new type?)
- promote to clickableLabel for the dark warning label to be able to reset the time-in-tip (same as warning label)
- Test time fields for times > 9:59 (I bet the text gets too wide)
- Test BREAK again
- Live mm:ss.zzz fields for dark intro/outro
- STOP and PLAY buttons can be a little larger
- "Playlist:" looks to be the default color, instead of same as "SD (..." color
- Reparent to flexible layouts, allow window resizing to affect widths of darkWaveformSlider, darkSongTitle, darkSongTable in X, and darkSongTableHeight and darkTreeView in Y
- Orange buttons in upper left for tabs/feature selection (like Mixxx)

**Comment by mpogue2** [comment]
TODO:
- Fix the clover-S situation to match the new locations of the Save menu options
- Provide a GUI affordance for the Beat/Measure options (Mixxx uses a magnet icon in the waveform display)

**Comment by mpogue2** [comment]
Initial population of the treeWidget: ee153781c54cd70b0f4dad159a0b46d5dc3a3f27

The treeWidget now populates itself with Tracks > type_names (one level), and Playlists > hierarchy_of_playlists.  You can use a flat musicDir/playlists folder, or it can be hierarchical (your choice, both ways work).

So, for example, my playlists are now organized like this:
- playlists
- - ECR
- - - 2022
- - - 2023
- - - etc
- - Jokers
- - - regular_playlist.csv
- - - 2022
- - - 2023

Initially, the Tracks and Playlists items are expanded to one level, for user convenience.

**Comment by mpogue2** [comment]
The "fake waveform" has been replaced with a real one, as of commit:  69d00eb238bce9f945fa0c439dc224e539161c78 .

For patter and singing calls, the waveform is now displayed correctly.  It's using a max() right now, maybe we should make that rms(), but it looks good for now.

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/440cadb7-895d-43ad-83ef-728e8eab7609)

**Comment by mpogue2** [comment]
Commit: 0a45b0ab19f2a7071f5687696439e66477f49c8d

It now displays all the songs now, and we can scroll through the darkSongTable list.  Nothing is hooked up yet though, so can't see playlists, can't double-click on songs, doesn't reflect filters, and changes to header widths are not persistent. 

BONUS: The new `darkLoadMusicList()` function is about twice as fast as the old `loadMusicList()`!  Old one: about 715ms to load the songTable from the pathStack.  New one: about 350ms to load the darkSongTable from the pathStack.

Note: Right now we're loading BOTH songTables at startup time.  Eventually, the DarkMode tab will replace the current Music tab, I think (hopefully with a Light Mode variant available). At that time, we'll save that ~365ms of startup time.

Current look now follows the user preferences for colors, too (instead of the prototype colors I was using before):
![image](https://github.com/mpogue2/SquareDesk/assets/6316003/ef7f63b3-f1ce-4a3a-850f-d00ccb59f218)

**Comment by mpogue2** [comment]
TODO: when I click on a song, immediately clear out the bgPixmap, so the old one is not showing while we're loading the new one.

**Comment by mpogue2** [comment]
Can now double-click in the darkSongTable, and song is properly loaded:  266c30ae316b31c1c30500947ce02ad9522bf0b0 .

**Comment by mpogue2** [comment]
The darkSearch and TreeView search features now work, and they are fast, as of: ee50233713f762a228b35b7cbb12796b37fb381b

darkSearch field combines the 3 old search fields together, like this:

type:label:title

If you just want to search on title, I've special cased it so that when there are no colons, it's just a title search.

So, for example:
- foo = looks for titles containing foo
- p:: = looks for patter
- s:riv = looks for Riverboat singers
- :riv = looks for all Riverboat songs
- p:esp:z = looks for patter songs from ESP where the title contains a Z
- #bar = the usual tag delimiters work, this looks for songs with the bar tag

STILL TODO: hitting ENTER to load a song in darkMode (I use this feature a lot)

**Comment by mpogue2** [comment]
TODO: Mixxx uses the hotkey ⌘F to put the focus into the search field.  I think that's a good idea for the DarkMode tab, and that key combination is not currently being used elsewhere.

**Comment by mpogue2** [comment]
Commit: 953624031e50176f242545fb341b247d205fef37

Wheel filter is now on again.  Can't use the mouse wheel to move the current position indicator in the waveform slider.  Wheel is still enabled for scrolling in the darkSongTable and the darkTreeView.

**Comment by mpogue2** [comment]
Commit: f37ff8e5f7369bbe44414ee9f81804bf125617bc

DarkSongTable items should not be editable.  They are not editable now (except for the # field, which isn't hooked up yet).

**Comment by mpogue2** [comment]
DarkMode seems very usable to me, but there are a couple of P0's that we need before we can really start wider user testing.

**TODOs before real user testing can begin (P0):**
- Fix ⌘S situation (this will bring it to parity with what we had before, just moved the functionality to separate menus)
- Hook up the intro/outro time boxes (beat/measure quantization works very very well, but it's not 100% sufficient)
- Playlists in the darkSongTable

**HIGH PRIORITY BUT COULD BE LATER (P1):**
- Allow width to change as window size is changed (can start testing without this)
- Figure out how to switch back/forth between Light and Dark modes (can start testing without this)
- Themes: a mechanism for customizing colors (external yaml file?  CSS file?  CSV file?) -- already have a way for expert users to customize SVG elements, by editing them with Inkscape (can start testing without this)
- Magnet icon (or something) for quantization? (maybe later, since these are menu items right now)

**Comment by mpogue2** [comment]
Commit: d4138f28c5f70378fa9ee3207d3d195527d5ace5

Can now set or change the intro/outro points on either the Music tab or the DarkMode tab.

**Comment by mpogue2** [comment]
Commit: 7bfb0b3a5f8fbb399321bf33022d75cfff987479

CurrentPos indicator is now hidden, until the first song has been loaded.

**Comment by mpogue2** [comment]
Commit: 31299bd725cb00abcaf4916a292c191df05ba093

**Resolution to "The ⌘S Situation"**
Because there are now separate menu items for saving Playlists, Cuesheets, and SD Sequences, we now move around the Cmd-S and Shift-Cmd-S shortcuts, based on what tab the user is on.  Note: can't currently SAVE something, unless you're on that tab.  We might want to change the isEnabled for these menu items to always enabled, if this turns out to be too unexpected (but, it's essentially the way the single File > Cmd-S and File > Shift-Cmd-S menu items worked before, when there was just the FIle ones and they changed names dynamically).

**Comment by mpogue2** [comment]
**NOTES on Mixxx playlists:**
- VIEW: Tracks and Playlist tracks are viewed separately, by clicking in the TreeVIew -- the views are mutually exclusive
- VIEW: can reorder columns by drag/drop
- VIEW: can turn columns on/off with context menu from the table header
- CREATE: Right click on "Playlists" > Create New Playlist gives dialog to provide a name, adds to Playlists in TreeView
- CREATE: also Right click on song > Create New Playlist
- CREATE: can also Duplicate an existing playlist
- LOAD PLAYLIST: Left Click on a playlist to load it (instantly)
- ADD: Right click on song > Add to Playlist > [single level list of playlists]
- ADD: Drag/drop to add track to playlist
- ADD: Playlists can have the same song more than once
- REORDER: Drag/drop to reorder tracks within a playlist (there is NO other way to reorder tracks in a playlist)
- REORDER: # field is NOT editable

**NOTES on current SquareDesk playlist UX:**
- VIEW: Tracks and Current Playlist tracks are in the same table at the same time
- VIEW: Can turn columns on/off via View > Columns > [list of columns]
- CREATE: Create a new playlist by loading an existing playlist and Save Playlist As...
- CREATE: or, can manually create a CSV file in the Playlist folder (it will be picked up at next app restart)
- LOAD PLAYLIST: Playlist > Load Playlist...
- LOAD PLAYLIST: Can do Playlist > Load Recent Playlist > [list of last 5 loaded]
- ADD/REORDER: menu items, context menu, and keyboard shortcuts to: add to top/bottom, move up/down, move to top/bottom, remove from playlist, clear playlist
- REORDER: can reorder by editing the # field

**Challenges:**
- I have ~100 playlists, one for each dance.  Mixxx handles a small # of playlists, but I have a large #, so a flat context menu with a few items is not gonna cut it.
- The darkTreeView currently has both Tracks (auto-populated with folder names/types, connected to search) and Playlists (auto-populated from CSV files in hierarchy in the playlists folder)

**Comment by mpogue2** [comment]
View of Rekordbox's palettes....the text is pretty darn small, but they manage to fit 4 current playlists (out of 12) AND the main songTable on the screen at one time:

<img width="1277" alt="image" src="https://github.com/mpogue2/SquareDesk/assets/6316003/7e028f29-482c-4ede-95d5-c3d597403f15">

Only the main songTable view has all the columns (about 30 to choose from, using a context menu, like Mixxx).  The 4 visible playlists just have artwork (also a preview column, that you can play the song from, by clicking on a tiny PLAY button that is over the artwork) + title.

SongTable columns can be reordered by dragging, like Mixxx.

The Playlist Palette can be hidden/shown via a dropdown on a small icon at middle right.

The search box on the right in the Collection section is the search box for the lower songTable.  This requires an extra row above the songTable, to hold the title "Collection", and a bunch of filter controls.

**Comment by mpogue2** [comment]
**TODO:**
- CMD-F should select all the text in the search box (so that typing new text will erase the old text)
- When ENTER is hit while in search box, where should focus go?

**Comment by mpogue2** [comment]
Scaling Phase 2: cc724e84d9a9835a60bf4844e3108b57f9e53bdf

I also tweaked the look to be a little more compact, and to accommodate the weird spacing rules imposed by the Qt layouts.  It now looks like this (fully compacted window size):

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/027a29b6-9e66-4c9a-a33f-03e5200db71d)


**TODO:**
- Make the waveform slider auto-scaling (horizontally) and remove the horizontal expanding spacer.
- BUG: Changing the pitch slider should update the darkTable's pitch column, too.  Same for Tempo.
- See if I can make the waveform peaks a little thinner, like Mixxx and the others (they look nicer than the current grey waveform peaks)

**Comment by mpogue2** [comment]
BUG: In Dark Mode (only), clicking on the darkSongbar waveform correctly moves the other 2 songBars handle, BUT clicking in the groove on the other 2 songBars does not correctly jump there (it kinda jumps randomly mostly in the right direction, but not always).  This is annoying.

**Comment by mpogue2** [comment]
Commit: 2ad2ea5ca4a1c5c1ea3b553e49bc31e94fd8e584

All of the widgets are resizing or repositioning themselves correctly now, INCLUDING the darkSeekBar.  Also added setFloatValue to the darkSeekBar, in case we want to update it more often than every second.  It no longer moves at irregular intervals, although depending on the slider size, it might move by irregular pixel amounts at every update (e.g. 1,1,1,2,1,1,1,2,etc. pixels moved once per second).

**Comment by mpogue2** [comment]
BUG kinda: Sparkly waveform on resize.  This is because I've only done the max() sampling of the waveform once (784 samples), and I'm just resampling in the integer domain (index into the array is an int).

Possible solutions:
- linear interpolation between adjacent waveform max() values
- just make the pixmap once, and tell it to rescale as it blits to the screen (that is, just resize the pixmap, rather than redrawing it)

**Comment by mpogue2** [comment]
**Current look when small:**

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/5df4fdc5-8342-4485-9efc-7616d7049e1d)

**Current look when window resized to be bigger (note that the waveform is scaling horizontally now):**

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/d918818a-c1de-42fd-8be1-06011332b89a)

**Comment by mpogue2** [comment]
TODO: allow track hierarchy?  e.g. patter has a subdirectory, and it appears as an item in the Tracks section of TreeView

**Comment by mpogue2** [comment]
Initial 3-column playlists: 4a05710c1e8460c0cded0102111c63096c7e9922

Now supports three sections that could hold playlists.  There is fake data in these, and they are not usable right now, but this is the general idea, and these changes are Phase 1 of adding playlists in.  This could also end up addressing part of #835.

This is still prototyping work.  The vertical splitter doesn't work quite right yet, for example.  And, I might or might not implement drag and drop to get stuff into these buckets (rekordbox calls them the 'palette').

**Comment by mpogue2** [comment]
The splitter between the darkSongtable and the playlist palette now works pretty well: cc0bb2c1a18132dc055f55d363a9b85b89b036de

The button to hide/show the playlists is currently now hidden (it was challenging to get it in there, so I've just hidden it for now, in case I want to bring it back later).

This is the basic layout for the playlist palette now, but I'm thinking it can also be used for search results (e.g. Patter, Singers, Rounds, etc.).  More UX work is needed to think through PRECISELY how to connect the search box, double clicking (in songTable vs palette box), and tempo/pitch controls to these new "palette" buckets (they are similar to rekordbox, but we have some added complexities).

Playlist palette CLOSED:
![image](https://github.com/mpogue2/SquareDesk/assets/6316003/c54b2c6d-1470-4d2c-a640-a35095b2028a)

Playlist palette OPEN:
![image](https://github.com/mpogue2/SquareDesk/assets/6316003/93ef485f-ac76-4232-bf3b-e6d6805b1c64)

Full-screen view (palette OPEN) on an M1 Macbook:
<img width="1429" alt="image" src="https://github.com/mpogue2/SquareDesk/assets/6316003/400b8c23-960a-4d34-ba2f-5ad4dd3fa255">

**Comment by mpogue2** [comment]
Feature?: One ESC to clear search, second ESC to stop playback?

**Comment by mpogue2** [comment]
BUG: When [ or ] are clicked or spin boxes are changed, need to update the waveform with the new colors (if singing call).

**Comment by mpogue2** [comment]
TODO: I'm not convinced that the per-beat Vamp is working precisely correctly in the darkWaveformSlider.  It works most of the time, but does it work ALL of the time?  Might be a pixel off?  This might affect Beat quantization more than Measure quantization.

TODO: Add a quantization control.  rekordbox uses a red "Q" that gets brighter when quantizing, and the quantization beat vs measure is somewhere in preferences (or rather, it claims to be -- I could not find it).
This should really be a tri-state control: none, beat, or measure.

**Comment by mpogue2** [comment]
Commit: e9e0db47b884146f7e7c74a7c1f3f1130e9e9297

New icon for palette slots.  I got rid of the prototype QToolButton on the right of the PlaylistNLabel, because I couldn't style it correctly (tried QPushButton and QToolButton, and I couldn't make it fade nicely into the background AND not cut off the icon on the left.).  I now just have a context menu attached to the whole label, rather than an icon on the right of each palette label.

The selections are now mutually exclusive between palette slot items and the {dark,}songTable items, in preparation for allowing ENTER to work (will consistently load the selected/highlighted song, regardless of which palette or songTable had the selection/highlighted entry).

Also, prototype (placeholder) context menu on the palette slot titles, in preparation for some (possibly) cool stuff soon.  I'll have to try these out, to see how usable these really are:

- [ ] Show Header (when checked, shows the header: #, Title, etc. in that slot, defaults to OFF for all)
- [ ] Select columns (checkboxes for Recent, Age, Pitch, Tempo -- separate controls for each palette slot, defaults to OFF for all.  Maybe no Recent or Age option at all for Playlists, since those are not IN the playlist's CSV file.)
- [ ] Delete on Load (when checked, double click or enter will load the song, and then delete the item from the palette slot.  This allows for making a queue that works similar to that in SqView, see #835.  Probably defaults to OFF for playlists, and Greyed out for Tracks.)
- [ ] Play on Load (when checked, will auto start playback on load via double-click or ENTER.  This is similar to the auto-play that we already have, but this is palette slot-specific.  Probably defaults to Off for all.)
- [ ] Auto-save (when checked, changes to a playlist (but not to track filter items) will be auto-saved to the CSV file, probably on by default for playlists, greyed out for tracks)
- [ ] Searchable (when checked, will pay attention to the filter in the Search field.  Probably on by default for track, off for playlists)

Because there will be a distinction between Playlists and Track filters in a palette slot, I'll need to distinguish by either label text (e.g. "Playlist: Jokers_2023.09.20" vs "Tracks: patter") or by color (?).  I'll think about this more.

**Comment by mpogue2** [comment]
Next short term:
- load a specific palette slot with a playlist or a track filter (context menus from the TreeWidget)
- right-click context menu with an "Add to" menu item (can add an item to a playlist in a palette slot this way), or maybe if there's only one Playlist in the Palette, right-click would mean add to the playlist (similar to SqView's queue)
- keyboard shortcuts for playlist add, move up/down/top/bot, remove should work on any playlist selection
- double click on a palette slot item, songTable item, or a darkSongTable item to load it

Later (maybe -- in addition to context menus and keyboard shortcuts):
- drag and drop items to add to playlists (similar to Rekordbox and Mixxx)
- drag and drop playlist items to reorder (similar to Rekordbox and Mixxx)
- could provide tool buttons in each playlist header to move up/down or option top/bot
- crates could be done with playlists (not much difference from playlists, other than not having a # column)

**Comment by mpogue2** [comment]
Some loading of playlist (only into slot 1 for now), with error handing now: 227676859701060b9c08de6c033678db7f39b538

Playlist items in the Palette can be double clicked to load them.

If there's an item in the playlist that has moved or changed name (and so it can't be found any more), it will be highlighted in red and text will be strikethrough.  Tooltip will suggest manually editing the .CSV file to fix it.  And, by RIGHT CLICKing on the playlist header, there will be two options shown:
- Edit playlist
- Reload playlist

The user can use RIGHT CLICK > Edit playlist to open TextEdit (on MacOSX) on the playlist file.  They then can File > Save the file, and come back to SquareDesk, where they can do RIGHT CLICK > "Reload playlist".  They should see the error go away, if the edit was successful.

Playlist items that are red (and therefore don't exist) don't do anything upon being double-clicked.  We might want to pop up an error dialog in the future, but for now I think the tooltip with manual edit options in the context menu will work OK.

<img width="1008" alt="image" src="https://github.com/mpogue2/SquareDesk/assets/6316003/b149385e-5fec-4870-805b-a48a075e6252">

**Comment by mpogue2** [comment]
Commit: 80f9a892b5d83ad20eeae992c091b444468f4143

SHIFT-CMD-{up, down, left, right} are now hooked up, if the selection is in the playlist1 slot.  Pressing these keys will move the selection item {up, down, to the top, to the bottom}.  Works only with slot #1 right now.

**TODO:**
- make shortcuts work with slots 2 and 3
- add in context menus for RIGHT CLICK over Playlist items
- factor the code (lots of duplicated lines right now)
- make it work with darkSongTable (gotta put in the # column first, to match the regular songTable)
- redo the menu item enablement code, so that these moves can be accessed from the menu, too
- put in ADD TO TOP/BOT functionality, e.g. from darkSongTable where # == ""
- should we also have an ADD AFTER CURRENT SELECTION option?  And/or ADD BEFORE CURRENT SELECTION?
- make ENTER work in slot 1
- make ENTER work in slots 2 and 3
- Should we have something like SHIFT-ENTER to act as ADD SONG TO BOTTOM OF SLOT 1?  That's almost the same as SqView's add-to-queue (but requires SHIFT -- this could be an option, if we really wanted to emulate SqView's UX)

**Comment by mpogue2** [comment]
Commit: 0d5ac2ebc33e659be19d4095e06d0a28ee871196

Load Playlist now (temporarily) loads the playlist into all 3 slots (0,1,2) for testing.  Double-click on any of the items in any slot OR press ENTER when a playlist slot item is selected to load that song for playback.  Keyboard shortcuts (SHIFT-CMD-{up/down/left/right/DEL) all work in slots 0,1,2 now.  Keyboard up/down also works in slots 0,1,2 now.  Code factoring is still needed here.

**TODO:**
- Slots should have type and label columns (but they should default to only showing the # and Title, unless the user changes it for that slot) .  Right now, the title is the whole title, e.g. "RIV 123 - foobar.mp3", which does not match the songTable or darkSongTable.  
- songType is a placeholder right now, until the above columns are added in.
- after a playlist song is loaded with ENTER, it's still selected but UP/DOWN don't work.  This is different from the songTable and darkSongTable, where UP/DOWN still work.  Gotta figure out how to best handle this (maybe focus should go back to the darkSearch field, and UP/DOWN should be enabled there?  I think that's the way the regular search fields work, IIRC...).

**Comment by mpogue2** [comment]
Commit: 16a033030b3acbfbbb31d020400a2f51c9077671

RIGHT-CLICK on a Playlist in the TreeWidget, and you can now show that playlist in (or "assign that playlist to", perhaps?) one of the three Palette slots.  The context menu in the Palette slot header now tells you which file you can Edit or Reload.  The Pallete slots now also have an additional context menu item to Remove from Palette.  

At startup, the last-loaded (via Playlist > Load) playlist will also be shown in Palette slot #1.  Finally, if there is no playlist in a slot, its title will be "Untitled Playlist", in anticipation of being able to add items to it soon.

NOTES: 
- Palette-loaded playlists are not yet savable or auto-savable, although they are editable (you can reorder items and delete items).  
- Can't add items to a playlist in DARKMODE yet.
- Playlists loaded into the slots are not persistent (once the app is restarted, the palette slots will be blank.)

**Comment by mpogue2** [comment]
TODO: double check power consumption  (idle, playing)

**Comment by mpogue2** [comment]
Commit: e043143a477d07032a8bf9ee922ff26661da931d

Double-clicking on a playlist in the TreeView will load that playlist into Palette Slot #1.  Double-clicking on a Track filter in the TreeView will just show that filter view in the darkSongTable.

**Comment by mpogue2** [comment]
TODO: When I click on a Playlist in the TreeWidget, it should populate the darkSongTable.  This has been done for Track filters, but not for the Playlists yet.

**Comment by mpogue2** [comment]
Commit: 614800743a7badab8e610eb47c3137f2aa505c88

RIGHT-CLICK on the darkSongTable to be able to add that song to the top/bottom of a playlist slot.  If there is anything in the playlist slot, RIGHT-CLICK on the playlist header will allow you to Save Playlist As, and you can save it to a CSV file.  

I also tested "round-tripping".  So, when an item is added to an "Untitled playlist", and the playlist is saved with RIGHT-CLICK Save Playlist As, the TreeWidget updates with the new playlist name, and RIGHT-CLICK on the playlist header allows for both Editing (in an external text editor) and Reloading (once external editing is complete).

Drag and drop is not yet supported (and I think Drag/Drop is probably easier to understand, although it requires a bit more dexterity with the mouse/trackpad).

**Comment by mpogue2** [comment]
Current pics:

Context menu (RIGHT_CLICK) over a darkSongTable item:
![image](https://github.com/mpogue2/SquareDesk/assets/6316003/ded4d1f3-eda7-442d-9648-e47ce33cc3d0)

Context menu (RIGHT_CLICK) over the playlist slot #1, that doesn't have a name yet:
![image](https://github.com/mpogue2/SquareDesk/assets/6316003/4d56bd6e-7cb6-461a-b0c7-b62337f79eeb)

Context menu (RIGHT_CLICK) over a playlist name in the TreeView:
![image](https://github.com/mpogue2/SquareDesk/assets/6316003/62cd1c52-60cd-4136-b715-d830395e3b71)

**Comment by mpogue2** [comment]
TODO: Should I split the TreeView into 2 TreeViews, the top one for Playlists, the bottom one for Tracks?
TODO: Should double-clicking a playlist or a track filter in the TreeView load it into the next EMPTY slot?

**Comment by mpogue2** [comment]
Commit: e8a2524a00c6ad41ec7e573c9cd5efed48aab240

RIGHT-CLICKing on a Track filter in the TreeView will pull up a context menu allowing the user to put that filter into a Palette Slot.  Items in the Track Pallete Slot can be double-clicked to load them.  

RIGHT-CLICKing on that slot label allows the user to Remove that filter from the slot.

**Comment by mpogue2** [comment]
Commit: a940c6c3a7363ecc3644e0032bedca3a6f442518

If you double-click on a Track Filter or a Playlist in the TreeView, it will load that item into the next empty palette slot.  If all the slots are full, it will load that item into the 3rd palette slot.

**Comment by mpogue2** [comment]
TODO: disallow the same playlist in more than one slot (if user does this, remove that playlist from the previous slot, and put the playlist where the user wants it).

**Comment by mpogue2** [comment]
**SAVE models:**

1) Auto-save:
PROS:
- works the same way the songTable itself works (instantly saved), so more consistent
- don't have to worry about saving at Quit time (Quit time always saves all changes, without needing to ask)
- works the way rekordbox does (which has lots of users), so some precedent here.
CONS:
- no way to revert (unless we make a .csv.bu when loaded, and delete it only at Quit time? I think that works...add Revert Changes would be added to the context menu)
- kicks off the Dropbox/Box.net/iCloud file watcher repeatedly on every auto-save (could limit this to once per minute, I suppose -- of course, the counterargument is that doing this in a database would also kick off the file watchers, so no real difference there, and actually our limiting it to once per minute is actually easier on the file watchers than multiple database updates)

2) Manual save:
PROS: 
- works the way the single playlist mechanism works today
CONS:
- hard to extend to having 3 playlists in various stages of being edited at the same time (e.g. Cmd-S is ambiguous, have to ask about up to 3 at Quit time)
- Well, it's manual.  So, at Quit time, we have to ask if there are unsaved changes.

**Tempo/Pitch models:**
1) Always affects songTable only
CONS: doesn't make sense. How does a playlist Tempo/Pitch get changed then?  By readding it?  nope.

2) Always affects only where it was loaded from (songTable, playlist)
PROS:
- Probably more intuitive
CONS:
- Requires some GUI affordance to tell us where it came from (use space under the songTitle, e.g. "from playlist: foobar", OR highlight the row of where it came from in Green, OR both ?)
- Some complexity if song was loaded from a TrackFilter slot, in which case it needs to change in both that slot (and any other slot that has that song) and the songTable. Because these are all the same song, really.  This is not so intuitive, but maybe I could get used to this....actually, the row would be highlighted in both places, so maybe that would make it more intuitive

3) Affects all instances of that song in any playlistTable or songTable
CONS:
- Doesn't make sense, because when editing we want to change it only in one place

----
**Hmmm...I think the winners are going to be:**

- SAVE: Auto-save with revert and once-per-minute limiting
- TEMPO/PITCH: affects where loaded from, with space under songTitle AND row highlighting (exactly one row in a single playlist, but potentially one row per slot in Track Filter slots)

I'll implement these to try them out, and see how they feel.

**Comment by mpogue2** [comment]
- Right now, I can populate more than one slot with the same playlist.  I think I should disallow that and go with the later choice.

**Comment by mpogue2** [comment]
Commit: d6995d4636be71fdc8570c217c7fa36b0d07c47e

Auto-save with 10-sec delay is now in.  I haven't implemented the revert stuff yet.

**Comment by mpogue2** [comment]
Commit: 9c0fc62bcf1e4568bfa0f1880ff59a9f195e5b6d

A playlist can only appear ONCE in the slot palette.  This prevents having to modify all the slots (potentially) when one slot is modified (and all the slots contain the same playlist), which seems like something people will not want to do anyway (it's awful confusing).  This sidesteps the complexity entirely.

**Comment by mpogue2** [comment]
Commit: 42a04c52178b36885a29b0ce069d54bf38267906

Hierarchical playlists should now work everywhere.  Playlists can now be in subfolders of the playlists folder, e.g. "musicDir/playlists/Jokers/2023/Jokers_2023.10.01", and everything should work.

Also, both playlists and track filters are now persistent in the palette slots.  If you load them up, and quit SquareDesk, they should be there again when you come back in.  Most people will probably keep around the ones that they are working on.

NOTE: As of a previous commit, playlists are AUTO-SAVED.  So, if you make changes to them, they'll be saved after 10 seconds of no editing.  There is no revert yet.

**Comment by mpogue2** [comment]
Bug: Need to disallow editing the track filters.  Right now the SHIFT-CMD-{up/down/left/right} keys will move the filter entries around, and that's incorrect.  Should be able to look at relPath[] to detect the track filters case, I would think.

**Comment by mpogue2** [comment]
TODO: Tempo/pitch changes need to modify something in the playlist and/or the darkSongTable.

**Comment by mpogue2** [comment]
Commit: 2d1326ace06d28ce9937187d5ce7e3338e9e1b58

Allows the user to delete a playlist item (item in a playlist in a palette slot) by pressing the SHIFT-CMD-DELETE key.  Changes will be auto-saved.  Also removed the red error message from the status bar, when in dark mode and a playlist is loaded that has errors (errors in this case are handled by highlighting the playlist item in red).

**Comment by mpogue2** [comment]
Commit: 5db82073d1f76997f4362207cc49483d70b8c274

Fixes a bug whereby the palette slot was not getting filled with the filtered song list at startup time (worked fine at double-click time).  Problem was that double-click is an extension of single-click, which filtered the darkSongTable.  At startup, this didn't happen, since there were no filters yet.  Fixed it so that track filtering by song type is correct at all times (including at startup), regardless of the state of the darkSongTable.

The main symptom was that I could double-click load "xtras" into a slot, but when I quit and came back in, the title of the slot was correct == "xtras", but the contents were basically the whole unfiltered list.

**Comment by mpogue2** [comment]
Commit: a4cdb817b32b0861ab407dc97bddf0f300dba3b1

If there's a Track Filter in one or more of the Palette Slots, they will all now be updated when the FileWatcher notices that files have changed in the musicDir.

**Comment by mpogue2** [comment]
SO MANY DETAILS.

But, I think I'm making good progress on all fronts.  All the playlist/track filter stuff seems to be in pretty good shape now.

Major missing item right now is being able to change a playlist item's pitch and tempo when a song is loaded from a playlist.  Track filter items are done, but playlist items are not, since it requires some thinking about the UX.  The same song can be in more than one playlist.  And, the same song can be on one playlist more than once.

Drag and drop would be nice, but can wait until later, since the context menu approach works for now.  It's a little clunky, but it will do, until I can figure out how to do DnD.  

The double-click-to-load-a-playlist/track filter is pretty intuitive, now that I've done it a couple of times.

**Comment by mpogue2** [comment]
Commit: 3b513033e5c38fe51de2af20066084fd7db047fe

If you want to edit the pitch/tempo of a playlist item, make sure that the playlist item is showing in one of the palette slots.  Then, double-click to load that item.  Change the pitch/tempo sliders, and just that one playlist item will be changed.  If you double-click from the songTable or darkSongTable, then changing pitch/tempo sliders will change it in the songTables, AND in any matching items in a Track Filter list, AND in the sqlite3 database (the way it works today) -- in this latter use case, the playlist items will NOT change.

**Comment by mpogue2** [comment]
Commit: 

Currently puts a "*** " before titles in the palette slots.  I'm not sure I like this, but I'm gonna start by testing this method.

This "editingArrow" is used to show the user which item is loaded, and is therefore the one changed when the Pitch and Tempo sliders are moved.  We might want to do **Bold** or _Italic_ or a color change or something else.  Just trying this out for now.

Hmmm...should the editingArrow also appear in the songTable and darkSongTable? or just the darkSongTable?

**Comment by mpogue2** [comment]
Commit: d15bd5af838e5cafed991b64f11b19de7fdb66e4

If a song is loaded from a Playlist or from the darkSongTable, a pencil icon will appear before the title that now has editable pitch and tempo, and the text will be bold/italic.  This will not happen to the darkSongTable, to preserve compatibility with the old songTable.  I don't know if this will make sense or not, I'm trying it.  In any case, the code is there to be able to make highlighting changes to QTableWidgetItems, and to remove that highlighting at the right time.

These highlighting changes will also not be made if the songTable is double-clicked.  Not sure that's the right approach, I'm just trying it to see how it feels.

**Comment by mpogue2** [comment]
Commit: a3e6e3466d91559ac4aa040f0e1486bed8347c06

I like the Black Rightwards Arrow (monochrome) somewhat better than the Pencil (which comes out multi-colored).

Before:

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/bb4eaf7f-9047-443a-8137-91a228447ec5)


After:

![image](https://github.com/mpogue2/SquareDesk/assets/6316003/eb8c6571-855a-4866-abfc-8ccf1fffb358)

**Comment by mpogue2** [comment]
It's looking pretty good so far, I think, and I have just a few items to work on next:

- Drag and Drop (big item)
  - tracks/playlists in TreeWidget into a palette slot (loads into a palette slot)
  - items up/down in palette slots (to move them -- maybe we don't do a delete gesture, since that's likely to be hit unintentionally)
  - items from palette slots to the playback section (to load)
- Organize the darkSongTable context menu better.  Right now, it has 6 items for copying to top/bot of palette slots 1/2/3.  They are currently not using hierarchical menus either, which makes 6 items instead of 2 * 3 items or 3 * 2 items.  
- Maybe key shortcuts for "Add to Bottom of palette slot #1/2/3"?  That's probably the most common operation.
- Populate the palette slot context menus.  Right now, just "Show in Enclosing Folder".  Probably need the usual up/down/top/bottom operations, along with shortcuts to help teach the user what to press.
- Maybe revert?  First view of a palette slot makes a copy, which the user can revert to via context menu
- Light mode, in addition to Dark mode (perhaps later?), and a way to switch between them
- .pro file commands to copy in the sliders, etc.  (it's PITA to do this by hand)
- fix the damn MacOS folder is a file problem? That's got to be a bug in the .pro file somewhere.  It's annoying but only hits when there's a new version of Qt, which is why it's lower priority.

**Comment by mpogue2** [comment]
Commit: ea93ca50a8d4414d6112755b6400bdcade88147e

The DarkMode tab now replaces the Music tab, when DARKMODE is #define'd in globaldefines.h.  I'm next going to work through each of the color and font problems, so that all tabs are in Dark Mode (always).

**Comment by mpogue2** [comment]
Bug:  when item in a playlist is double-clicked, it loads with the full name (including label and dash), instead of just the title.
Bug: When resizing the mainWindow, the singing call highlighting end position moves all over the place and is frequently wrong.

**Comment by mpogue2** [comment]
Commit: 5012e84ca453c159942b354ae373044cf4c833ed

When a singing call is playing, I really want to know what the NEXT section is.  Now when playing, we anticipate the section change by about 2.5 seconds, giving the UX time to display it slightly before it's needed.

The UX only updates about 1X/sec, so by anticipating by 2.5 sec, we see the section change about 1 - 1.5s before it's needed.  This anticipation only happens when music is playing.  When it's NOT playing, it works as before (cursor position indicates section precisely).

**Comment by mpogue2** [comment]
Commit: 22c947f61947326fe0b14c1d130e3cd3c0086d62

Tag editing in the darkSongTable is now restored.  I also tweaked the Edit Tags dialog a bit to make it easier to read.

**Comment by mpogue2** [comment]
Commit: 9eae62d90ce9b97dedcd99e68e6615687ea6a27b

Now that the (new) DarkMusic tab can replace the (old) Music tab, we can name it just "Music".

**Comment by mpogue2** [comment]
Commit: ca86b83068ad64c12c1766e4a3f494023c24e316

In preparation for release, bump all version numbers to 1.0.6 .  Note also that the About box has changed, because Qt6.5 doesn't implement rich text (including links) now.  See https://bugreports.qt.io/browse/QTBUG-116757 .

Note that DARKMODE is now the default.  I'll work on making it switchable (Dark/Light) at a later date.

**Comment by mpogue2** [comment]
Things missing:
- Slidable separator between playlists
- Ability to drag from one playlist and drop on another
- rename a playlist (e.g. load pl1, rename to pl2, then delete/move around items)
- auto-tags that tell me not just what singer level is, but what all the singer levels are

Crash:
- add anything to pl2, then d.c. something in pl1 (boom)

**Comment by mpogue2** [comment]
Moving the open items to a different Issue.  Closing this one, because DarkMode is now implemented and usable.
